src/main/java/com/appland/appmap/record/RecordingSessionGeneric.java

```diff
@@ -2,44 +2,57 @@
 
 import com.appland.appmap.output.v1.CodeObject;
 import com.appland.appmap.output.v1.Event;
-import com.appland.appmap.util.Logger;
 
-import java.util.Vector;
 import java.util.HashSet;
+import java.util.Optional;
+import java.util.Vector;
 
 public class RecordingSessionGeneric implements IRecordingSession {
-  protected Vector<Event> events = new Vector<Event>();
-  protected HashSet<String> classReferences = new HashSet<>();
-
-  public synchronized void add(Event event) {
-    this.events.add(event);
-
-    final String key = event.definedClass +
-      ":" + event.methodId +
-      ":" + event.isStatic + 
-      ":" + event.lineNumber;
-    this.classReferences.add(key);
-  }
-
-  public void start() {
-    throw new UnsupportedOperationException();
-  }
-
-  public String stop() {
-    throw new UnsupportedOperationException();
-  }
-
-  protected CodeObjectTree getClassMap() {
-    CodeObjectTree registeredObjects = Recorder.getInstance().getRegisteredObjects();
-    CodeObjectTree classMap = new CodeObjectTree();
-    for (String key: this.classReferences) {
-      String[] parts = key.split(":");
-     
-      CodeObject methodBranch = registeredObjects.getMethodBranch(parts[0], parts[1], Boolean.valueOf(parts[2]), Integer.valueOf(parts[3]));
-      if (methodBranch != null)
-        classMap.add(methodBranch);
+    protected Vector<Event> events = new Vector<Event>();
+    protected HashSet<String> classReferences = new HashSet<>();
+
+    public synchronized void add(Event event) {
+        this.events.add(event);
+
+        if (event.parentId == null) {
+            this.classReferences.add(
+                    event.definedClass +
+                            ":" + event.methodId +
+                            ":" + event.isStatic +
+                            ":" + event.lineNumber);
+        } else {
+          Optional.ofNullable(events.stream()
+                    .filter(it -> it.id == event.parentId)
+                    .findFirst()
+                    .orElseThrow(() -> new IllegalArgumentException("Missing parent Event.")))
+                  .ifPresent( parentEvent ->
+                          this.classReferences.add(parentEvent.definedClass +
+                                  ":" + parentEvent.methodId +
+                                  ":" + parentEvent.isStatic +
+                                  ":" + parentEvent.lineNumber)
```

Comment:
The parent event would have already been through the first `if` block. The class reference should already be added, so this can safely be removed. It's also dangerous for us to throw an unhandled exception in this section of code as it can trigger undefined behavior in the user's application.

Reply:
mmm very interesting, I will take a look tomorrow

Reply:
@dustinbyrne after checking removing this cause some integration test to fail, but I did remove the exception throwing.